### PR TITLE
Feature/997 Perform a staged migration of Stages & Statuses data

### LINKED ISSUE #997: Perform a staged migration of Stages & Statuses data
**Background**
Now that our new set of stages and statuses have been confirmed we should migrate all `exercises` and `applicationRecords` from the old stages & statuses to the new equivalents.

**Feature(s)**
We will perform a 3-stage migration: 1. Test migration of selected archived exercises; 2. Full migration of remaining archived exercises; 3. Migration of live exercises.

1. Write migration script
2. Test by migrating the following archived exercises only: 9, 15, 37, 66, 82,
3. PAUSE and check exercise stats before and after on the above exercises 
4. If satisfactory, migrate all archived exercises
5. Schedule the migration of live exercises to take place out of working hours, e.g. weekend

### PR DESCRIPTION
**Pull request by HalcyonJAC**
Closes #997.

- Add a node script `updateApplicationRecordStageStatus` to migrate the stage and status in the `applicationRecord` document with the given `exerciseId` and `version`.
- Add a callable cloud function `updateApplicationRecordStageStatus` to update the stage and status in the `applicationRecord` document with the given `exerciseId` and `version`.
- The previous stage and status will be stored to enable rollback.

  applicationRecord:
  ```
  _backups: {
    processingVersion1: {
      stage: '',
      stageLog: {},
      status: '',
      statusLog: {},
    }
  }
  ```

  exercise:
  ```
  _backups: {
    processingVersion1: {
      _applicationRecords: {},
    }
  }
  ```

Note: This PR is related to [admin: #2294 Use agreed set of v2 statuses](https://github.com/jac-uk/admin/pull/2325).

Screenshot of the script output log:

Screenshot 2024-02-21 at 10 37 16

### CODE REVIEW

**Comment by warrensearle** [comment]
@HalcyonJAC how much effort would it be write a migration script to go back the other way - from v2 stages and statuses back to v1 stages and statuses?

**Comment by warrensearle** [comment]
As discussed this morning @HalcyonJAC please can you create a cloud function wrapper for the migration script, and also enable roll-back of the changes (store the original statuses in the `applicationRecord` document).

This will then be called on a per-exercise basis from the admin UI.

File: functions/actions/applicationRecords/updateApplicationRecordStageStatus.js [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+   * Updates specified applications records with the new stage and status
+   * @param {*} `params` is an object containing
+   *   `exerciseId` (required) ID of exercise
+   *   `version`    (required) _processingVersion of the exercise
+   */
+  async function updateApplicationRecordStageStatus(params) {
+    const { exerciseId, version } = params;
+
+    // get application records from reference numbers
+    const applicationRecords = await getDocuments(db.collection('applicationRecords').where('exercise.id', '==', exerciseId));
```
Comment (warrensearle) [inline functions/actions/applicationRecords/updateApplicationRecordStageStatus.js:23]: We could speed this up slightly, particularly when run via nodescript, by adding `.select('_backup', 'stage', 'status')` to the query. So the line becomes something like:
```
    const applicationRecords = await getDocuments(db.collection('applicationRecords').where('exercise.id', '==', exerciseId).select('_backup', 'stage', 'status'));
```

**Review by warrensearle — CHANGES_REQUESTED** [review]
This looks great @HalcyonJAC. Your code is always really nicely written and easy to follow.

We do need to update the stages. I'll send across proposal mappings for those asap

👍

File: functions/actions/applicationRecords/updateApplicationRecordStageStatus.js (hunk 1 above)
Comment (HalcyonJAC) [reply functions/actions/applicationRecords/updateApplicationRecordStageStatus.js]: Thank you for the suggestion 👍

**Review by drieJAC — APPROVED** [review]

**Review by warrensearle — APPROVED** [review]
Great to see some unit tests in there 😃

**Pull request merged**